Change strategy for Transaction Submission to Espresso

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Modifies the processing and Handling of Transaction Submission / Verification to Espresso.  The point / design of this change is to provide feature / implementation parity with the current transaction submission strategy, but without the overhead of spawning an unbounded number of goroutines.

In order to achieve this, this PR moves to a job queue / worker implementation where there is a bounded number of workers for processing jobs, and a bounded number of scheduelers. Each of these processing points spawns a goroutine, but we have them limited to a bounded number of goroutines that will stay constant rather than scaling with the number of requests.

### This PR does not:
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->
* Currently add a limit on the number of attempts jobs are attempted
* inspect error cases to make an informed decision about whether to reattempt or not
* Currently clean up all channels appropriately
* Currently pause submission when the Batcher is Stopped

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210192466975806

Ayiga committed:
Add worker queue transaction submission to Espresso

The Workflow for submitting transactions to Espresso currently will
spawn an unbounded number of goroutines to submit transactions to
Espresso.  This can be incredibly resource heavy, and may end up
actually harming performance due to schedule thrashing.

This change adds a basic worker based implementation for processing
individual job requests for transaction submission and receipt verification.
This should help limit the total number of requests that are being
submitted to Espresso at any given time.

There are improvements that are able to be made to this, but this serves
as a basic implementation that matches what was happening originally
with one delay missing.

Ayiga committed:
Add 100ms delay in repeat transaction verification

When a transaction fails to verify, it will be added into the job queue and
reattempted without any delay between these attempts.  This can end up
with a potential DOS vector for multiple requests being attempted back
to back as soon as they are able to be performed.  There should be a
delay between these attempts, ideally without impacting other requests
that don't need to wait from continuing.

Add a 100ms delay to submission verification re-attempts.  This is not as
optimal as it could be and will prevent other jobs from coming in at the
moment.

Modify Espresso Submitter to respect shutdown

The Espresso Submitter should stop processing things when it is told
to do so by the Batcher.  The shutdownCtx governs whether things should
run or not.

The stop worker path has been modified to just utilize the same
WaitGroup and context as the other processes.

philippecamacho commented on op-batcher/batcher/driver.go:
```diff
@@ -215,6 +216,14 @@ func (l *BatchSubmitter) StartBatchSubmitting() error {
+			WithWaitGroup(l.wg),
+			WithEspressoClient(l.Espresso),
+		)
+		l.submitter.SpawnWorkers(4, 4)
```
Would it be worth to create some configuration variable for these values? Is there any advantage to have different numbers here, that is why not use the same number of SubmitTransaction/VerifyReceipt  workers?

Ayiga replied:
I originally had the number of workers match across both.  I suppose having them separate allows for you to tune these separately.  This can be beneficial, especially when it comes to the verify receipt workers, as they can end up paused and asleep when re-attempting  a job that hasn't been verified yet.  So in that case you might want more receipt verify workers in order to try and ensure that we are are still making some progress even while a few of our workers are sleeping.

philippecamacho replied:
I see, thanks.

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+// configure the EspressoTransactionSubmitterConfig.
+type EspressoTransactionSubmitterOption func(*EspressoTransactionSubmitterConfig)
+
+// WithEspressoClient is an option that can be used to set the Espresso client
```
Should we replace `WithEspressoClient`  with `WithContext`?

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+// configure the EspressoTransactionSubmitterConfig.
+type EspressoTransactionSubmitterOption func(*EspressoTransactionSubmitterConfig)
+
+// WithEspressoClient is an option that can be used to set the Espresso client
```
```suggestion
// WithContext is an option that can be used to set the Espresso client
```

Ayiga replied:
Addressed in db23892adc18ad7fe5ff7c51cfffb1203b8cb3ac.

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+}
+
+// NewEspressoTransactionSubmitter creates a new EspressoTransactionSubmitter
+// with the given context and espresso client.  It will create a new
```
> It will create a new

Looks like the sentence is not finished.

Ayiga replied:
Addressed in db23892adc18ad7fe5ff7c51cfffb1203b8cb3ac.

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+	config := EspressoTransactionSubmitterConfig{
+		Ctx:                                context.Background(),
+		Wg:                                 new(sync.WaitGroup),
+		SubmitJobQueueCapacity:             1024,
```
Why do we have this difference of size (1024 v/s 10)?

Ayiga replied:
This size difference is based on where I felt the bottle necks would occur.  They just "felt" right, and I don't have any other reasoning beyond that.  That being said, we can absolutely adjust them if we'd like, and motivate their values with better data / instincts.

philippecamacho replied:
Sounds good.

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+// SubmitTransaction will submit a transaction to the Job queue.
+//
+// NOTE: This submits to a channel, and as a result, if the channel is full,
+// it will block execution until the channel is able to accept the job.
```
Can we detect when this happens to log some message?

Ayiga replied:
We could, but it could make the code more difficult to read / reason about:

```go
func (s *espressoTransactionSubmitter) SubmitTransaction(job *espressoCommon.Transaction) {
	submit := espressoSubmitTransactionJob{
		transaction: job,
	}

	select {
	case s.submitJobQueue <- submit:
		return
	default:
		// Channel is full.
		s.log.Warn("Submit transaction job queue is full, will block submitting job", "job", hexutil.Encode(crypto.Keccak256(job.Data)))
	}

	// Blocked Send
	s.submitJobQueue <- submit
}
```

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+			}
+		}
+
+		// TODO: Evaluate the specific error type, and determine if we
```
Are we ok leaving this TODO for after the audit?

Ayiga replied:
I am alright with this.  I was just attempting to feature match what @QuentinI had already done.

philippecamacho commented on op-batcher/batcher/espresso.go:
```diff
@@ -21,43 +22,529 @@ import (
+		}
+
+		if jobAttempt.job.attempts > 3 {
+			// We have submitted this job quite a bit.
```
Is this part of some TODO?

Ayiga replied:
Initially intended to be so, yes.  But perhaps this is best removed for now.

Ayiga replied:
Addressed in db23892adc18ad7fe5ff7c51cfffb1203b8cb3ac.

Ayiga committed:
Update comments and code based on feedback

The feedback from @philippecamacho has pointed out some
implementation details, and comments that were lacking, unfinished, or
lacked sufficient clarity.  These issues have been addressed where
immediately obvious, and not requiring further clarification.

philippecamacho reviewed (approved):
LGTM
